README.md
# Juttle Twitter Adapter

Twitter adapter for the [Juttle data flow
language](https://github.com/juttle/juttle).

## Examples

Filter the real-time Twitter stream for all tweets containing "potus"

```juttle
read twitter -to :end: 'potus'
```

Read the last 100 historical tweets referencing "potus":

```
read twitter -from :0: -limit 100 'potus'
```

An end-to-end example is described [here](https://github.com/juttle/juttle-engine/tree/master/examples/twitter-race) and deployed to the demo system [demo.juttle.io](http://demo.juttle.io/?path=/examples/twitter-race/twitter.juttle).

## Installation

Like Juttle itself, the adapter is installed as a npm package. Both Juttle and
the adapter need to be installed side-by-side:

```bash
$ npm install juttle
$ npm install juttle-twitter-adapter
```

## Configuration

The adapter needs to be registered and configured so that it can be used from
within Juttle. To do so, add the following to your `~/.juttle/config.json` file:

```json
{
    "adapters": {
        "twitter": {
            "consumer_key": "...",
            "consumer_secret": "...",
            "access_token_key": "...",
            "access_token_secret": "..."
        }
    }
}
```

To obtain Twitter credentials, first set up a [Twitter App](https://apps.twitter.com/) and create an OAuth token.

## Usage

The adapter can run in two modes -- streaming or historical, controlled by the `-from` and `-to` options.

For historical searches, specify `-from :0:` (and optionally `-to :now:`) to indicate that the read should start at the beginning of time and continue to the present.

For live searches, specify `-to :end:` (and optionally `-from :now:`) to indicate that the read should start at the current time and continue indefinitely.

In both cases the adapter requires a single search term in the filter expression, as shown in the examples above.

Historical searches are somewhat constrained by the fact that Twitter's API returns
tweets in reverse chronological order, but Juttle semantics require data points
to be emitted in order.

To handle this, the adapter buffers all the points in memory before emitting
them down the flowgraph. There is a limit to control how many points are
buffered, which defaults to 1000. This can be overridden using the `limit`
option.

In streaming mode, the adapter will buffer all points for a configurable delay (the `lag` option)
so that they can be properly sorted into increasing time order.

### Options

Name | Type | Required | Description
-----|------|----------|-------------
`from`  | moment | no | time to start reading the data -- must be either :now: or :0:
`to`  | moment | no | time to stop reading the data -- must be either :now: or :end:
`limit`  | integer | no | maximum number of tweets to emit in historical mode (default: 1000)
`fetchSize`  | integer | no | number of tweets to fetch per round-trip in historical mode (default: 100, max: 100)
`lag`  | duration | no | in streaming mode, delay for the given amount of time to reorder incoming points (default: 2 seconds)

## Contributing

Want to contribute? Awesome! Don’t hesitate to file an issue or open a pull
request.


CHANGELOG.md
# Change Log

This file documents all notable changes to Juttle. The release numbering uses [semantic versioning](http://semver.org).

## 0.4.0

### Major Changes

- Updated to be compatible with Juttle 0.7.0.

## 0.3.1

### Bug Fixes

- Fixed a misuse of the promise library that caused all historical reads to fail. [#13]

## 0.3.0

### Major Changes

- Updated to be compatible with Juttle 0.5.x, including options change so that whether to do streaming or historical is controlled by the `-from` and `-to` combination and not `-streaming <true/false>`. [#12]
- Changed the `-count` option to `-limit` and the `-delay` option to `-lag` to be consistent with other adapters. [#12]

## 0.2.2

### Bug Fixes

- Updated the README to reflect the proper way to configure the adapter.

## 0.2.1

### Bug Fixes

- Relaxed the peerDependency on Juttle to be >= 0.2.0.

## 0.2.0

### Major Changes

- Updated to be compatible with Juttle 0.2.0.

## 0.1.0

### Major Changes

- Initial version


lib/twitter-backend.js
var Juttle = require('juttle/lib/runtime').Juttle;
var JuttleMoment = require('juttle/lib/moment/juttle-moment');
var Twitter = require('twitter');
var Heap = require('heap');

var client;

var Read = Juttle.proc.source.extend({
    sourceType: 'batch',
    procName: 'readx-twitter',

    initialize: function(options, params, pname, location, program, juttle) {
        this.logger.debug('intitialize', options, params);

        // XXX hack
        if (! (params.filter_ast && params.filter_ast.type === 'SimpleFilterTerm')) {
            throw new Error('need a filter');
        }
        this.query = params.filter_ast.expression.value;

        this.do_streaming = options.stream || false;
        this.total = options.total || 1000;
        this.count = options.count || 100;

        this.delay = options.delay || JuttleMoment.duration(1, 's');

        // buffer of received tweets for search mode
        this.buffer = [];

        // live buffer for tweets in streaming mode
        this.live = new Heap(function(p1, p2) {
            return p1.time.lt(p2.time) ? -1 : 1;
        });
    },

    start: function() {
        if (this.do_streaming) {
            this.start_streaming();
        } else {
            this.search();
        }
    },

    start_streaming: function() {
        var self = this;

        this.live_window_start = new JuttleMoment().add(this.delay);
        this.live_window_next = this.live_window_start.add(this.delay);

        this.program.scheduler.schedule(this.live_window_next.unixms(), function() { self.stream_points() });

        client.stream('statuses/filter', {track: this.query}, function(stream) {
            self.stream = stream;

            stream.on('data', function(tweet) {
                var pt = self.toJuttle(tweet);
                if (pt) {
                    if (pt.time.gt(self.live_window_start)) {
                        self.logger.warn('point arrived late', pt.time.toJSON(), self.live_window_start.toJSON());
                        return;
                    }
                    self.live.push(pt);
                    self.logger.debug('buffered', self.live.size(), 'pts')
                }
            });

            stream.on('error', function(error) {
                // XXX how to handle errors
                self.logger.error(error)
            });
        });
    },

    stream_points: function() {
        var self = this;

        this.logger.debug('in stream', this.live_window_next.toJSON());

        var tosend = [];
        this.live_window_start = this.live_window_start.add(this.delay);
        while (this.live.size() && (this.live.peek().time < this.live_window_start)) {
            tosend.push(this.live.pop());
        }
        if (tosend.length !== 0) {
            this.logger.debug('emitting', tosend.length, 'points', JSON.stringify(tosend, null, 4));
            this.emit(tosend);
        } else {
            this.emit_tick(this.live_window_start);
        }

        this.live_window_next = this.live_window_next.add(this.delay);
        this.program.scheduler.schedule(this.live_window_next.unixms(), function() { self.stream_points(); });
    },

    search: function() {
        var self = this;

        var opts = {
            q: this.query,
            count: Math.min(this.count, this.total - this.buffer.length)
        };

        if (self.max_id) {
            opts.max_id = self.max_id;
            opts.count += 1;
        }

        this.logger.debug('search buffered=' + this.buffer.length, 'max_id=' + opts.max_id, 'count=' + opts.count);
        client.get('search/tweets', opts, function(err, tweets, response) {
            if (err) {
                return self.logger.error(err);
            }

            var statuses = tweets.statuses;

            // Yuck! Twitter's API returns the tweet that matches the max_id
            // instead of making it -1, and Javascript can't do 64 bit math, so
            // ignore the first tweet that arrived if this was a max_id query.
            //
            // It also pages in reverse order so we need to buffer all the
            // results up to the configured max.
            if (self.max_id) {
                statuses = statuses.slice(1);
            }

            var done = true;
            if (statuses.length !== 0) {
                self.buffer = self.buffer.concat(statuses);
                self.logger.debug('buffered', statuses.length, 'tweets', 'count=' + self.buffer.length)
                self.max_id = statuses[statuses.length - 1].id;
                done = false;
            }

            if (!done && (self.buffer.length >= self.total)) {
                self.logger.warn('reached limit of', self.total, 'tweets... stopping search');
                done = true;
            }

            if (done) {
                self.logger.debug('all done... emitting', self.buffer.length, 'tweets');
                self.emit_tweets(self.buffer);
                self.eof();
            } else {
                self.search();
            }
        });
    },

    teardown: function() {
        this.logger.debug('in teardown: stream', this.stream ? 'active' : 'not active');
        if (this.stream) {
            this.stream.destroy();
        }
    },

    toJuttle: function(tweet) {
        if (tweet.created_at === undefined) {
            return null;
        }

        return {
            time: new JuttleMoment({rawDate: new Date(tweet.created_at)}),
            user: '@' + tweet.user.screen_name,
            text: tweet.text
        };
    },

    // Emit the array of tweets in reverse order since that's how they come in.
    emit_tweets: function(tweets) {
        var pts = [];

        for (var i = tweets.length - 1; i >= 0; --i) {
            var pt = this.toJuttle(tweets[i]);
            if (pt) {
                pts.push(pt);
            }
        }

        this.emit(pts);
    }
});

var TwitterBackend = function(config) {
    // Just pass the config directly to the twitter client.
    client = new Twitter(config);

    return {
        name: 'twitter',
        read: Read
    };
};

module.exports = TwitterBackend;


lib/twitter-adapter.js
'use strict';

/* global JuttleAdapterAPI */
let AdapterRead = JuttleAdapterAPI.AdapterRead;
let JuttleMoment = JuttleAdapterAPI.types.JuttleMoment;
let StaticFilterCompiler = JuttleAdapterAPI.compiler.StaticFilterCompiler;
let Twitter = require('twitter');
let Heap = require('heap');
let Promise = require('bluebird');

let config;

class TwitterFilterCompiler extends StaticFilterCompiler {
    visitFulltextFilterTerm(node) {
        return node.text;
    }
}

class Read extends AdapterRead {
    static get timeRequired() { return true; }

    constructor(options, params) {
        super(options, params);

        if (!params.filter_ast) {
            throw this.compileError('FILTER-FEATURE-NOT-SUPPORTED', {
                feature: 'read with no filter'
            });
        }

        let filterCompiler = new TwitterFilterCompiler();
        this.query = filterCompiler.visit(params.filter_ast);

        let from = this.options.from || this.defaultTimeOptions().from;
        let to   = this.options.to   || this.defaultTimeOptions().to;

        // For now the adapter only accepts the following time modes:
        //
        //    -from :0:    (-to :now:)  (pure historical)
        //   (-from :now:)  -to :end:   (pure live)
        //
        if (from.unixms() === 0 && to.eq(params.now)) {
            this.streaming = false;
        } else if (from.eq(params.now) && to.isEnd()){
            this.streaming = true;
        } else {
            throw this.compileError('INVALID-OPTION-COMBINATION', {
                option: 'from/to',
                rule: 'pure historical or pure live'
            });
        }

        this.logger.debug('initializing in', this.streaming ? 'streaming' : 'historical', 'mode');

        // Historical read limit
        this.limit = options.limit || 1000;

        // Number to fetch for each historical read
        this.fetchSize = options.fetchSize || 100;

        // Live buffer size
        this.bufferLimit = options.bufferLimit || 10000;

        // buffer of received tweets for search mode
        this.buffer = [];

        // live buffer for tweets in streaming mode
        this.reorderBuffer = new Heap(function(p1, p2) {
            return p1.time.lt(p2.time) ? -1 : 1;
        });

        // points arriving before this time are dropped
        this.reorderFrom = params.now.quantize(JuttleMoment.duration(1, 's'));

        this.warnLate = this.throttledWarning('POINTS-ARRIVED-LATE', {
            proc: 'read twitter'
        });

        this.warnOverflow = this.throttledWarning('BUFFER-LIMIT-EXCEEDED', {
            buffer: 'read twitter buffer'
        });

        this.client = Promise.promisifyAll(new Twitter(config));
    }

    static allowedOptions() {
        return AdapterRead.commonOptions().concat(['limit', 'fetchSize', 'bufferLimit']);
    }

    defaultTimeOptions() {
        return {
            from: this.params.now,
            to: this.params.now,
            lag: JuttleMoment.duration(2, 's')
        };
    }

    periodicLiveRead() {
        return true;
    }

    start() {
        if (this.streaming) {
            this._startStreaming();
        }
    }

    read(from, to, limit, state) {
        if (this.streaming) {
            return this._fetchLive(from, to, limit, state);
        } else {
            return this._fetchHistorical();
        }
    }

    _startStreaming() {
        this.logger.debug('starting streaming');
        this.client.stream('statuses/filter', {track: this.query}, (stream) => {
            this.stream = stream;

            stream.on('data', (tweet) => {
                let pt = this.toJuttle(tweet);
                if (pt) {
                    if (pt.time.lt(this.reorderFrom)) {
                        this.warnLate();
                    } else if (this.reorderBuffer.size() >= this.bufferLimit) {
                        this.warnOverflow();
                    } else {
                        this.reorderBuffer.push(pt);
                        this.logger.debug('buffered', this.reorderBuffer.size(), 'pts');
                    }
                }
            });

            stream.on('error', (error) => {
                // XXX how to handle errors
                this.logger.error(error);
            });
        });
    }

    _fetchLive(from, to) {
        this.logger.debug('_fetchLive', from.toJSON(), '-', to.toJSON(),
                          'bufferSize', this.reorderBuffer.size());

        // XXX need to handle limit
        let points = [];
        while (this.reorderBuffer.size() && (this.reorderBuffer.peek().time < to)) {
            points.push(this.reorderBuffer.pop());
        }
        this.reorderFrom = to;
        this.logger.debug('returning', points.length, 'points');
        return Promise.resolve({
            points: points,
            readEnd: to
        });
    }

    // Perform the historical search. Returns a promise when done.
    _fetchHistorical() {
        let opts = {
            q: this.query,
            count: Math.min(this.fetchSize, this.limit - this.buffer.length)
        };

        if (this.max_id) {
            opts.max_id = this.max_id;
            opts.count += 1;
        }

        this.logger.debug('search buffered=' + this.buffer.length,
                          'max_id=' + opts.max_id, 'count=' + opts.count);

        return this.client.getAsync('search/tweets', opts)
        .then((tweets) => {
            let statuses = tweets.statuses;
            this.logger.debug('search returned', statuses.length, 'tweets');

            // Yuck! Twitter's API returns the tweet that matches the max_id
            // instead of making it -1, and Javascript can't do 64 bit math, so
            // ignore the first tweet that arrived if this was a max_id query.
            //
            // It also pages in reverse order so we need to buffer all the
            // results up to the configured max.
            if (this.max_id) {
                statuses = statuses.slice(1);
            }

            let done = true;
            if (statuses.length !== 0) {
                this.buffer = this.buffer.concat(statuses);
                this.logger.debug('buffered', statuses.length, 'tweets',
                                  'count=' + this.buffer.length);

                this.max_id = statuses[statuses.length - 1].id;
                done = false;
            }

            if (!done && (this.buffer.length >= this.limit)) {
                this.logger.warn('reached limit of', this.limit,
                                 'tweets... stopping search');
                done = true;
            }

            if (done) {
                this.logger.debug('all done... emitting', this.buffer.length, 'tweets');
                let points = [];
                for (let i = this.buffer.length - 1; i >= 0; --i) {
                    points.push(this.toJuttle(this.buffer[i]));
                }
                return {
                    points: points,
                    eof: true
                };
            } else {
                return this._fetchHistorical();
            }
        });
    }

    teardown() {
        this.logger.debug('in teardown: stream', this.stream ? 'active' : 'not active');
        if (this.stream) {
            this.stream.destroy();
        }
    }

    toJuttle(tweet) {
        if (tweet.created_at === undefined) {
            return null;
        }

        return {
            time: new JuttleMoment({rawDate: new Date(tweet.created_at)}),
            user: '@' + tweet.user.screen_name,
            text: tweet.text
        };
    }
}

let TwitterAdapter = function(cfg) {
    config = cfg;

    return {
        name: 'twitter',
        read: Read
    };
};

module.exports = TwitterAdapter;
